firebase-storage/src/main/java/com/google/firebase/storage/StorageReference.java

```diff
@@ -563,14 +566,122 @@ public StreamDownloadTask getStream(@NonNull StreamDownloadTask.StreamProcessor
+   * next set of items (files) and prefixes (folders) under this StorageReference.
+   *
+   * <p>"/" is treated as a path delimiter. Firebase Storage does not support invalid object paths
+   * that end with "/" or contain two consecutive "//". All invalid objects in GCS will be filtered.
+   *
+   * @param maxResults The maximum number of results to return in a single page. Must be greater
+   *     than 0 and at most 1000.
+   * @param pageToken A page token from a previous call to list.
+   * @return A a {@link Task} that returns the next items and prefixes under the current
+   *     StorageReference.
+   */
+  @NonNull
+  @PublicApi
+  public Task<ListResult> list(int maxResults, @NonNull String pageToken) {
+    Preconditions.checkArgument(maxResults > 0, "maxResults must be greater than zero");
+    Preconditions.checkArgument(maxResults <= 1000, "maxResults must be at most 1000");
+    Preconditions.checkArgument(
+        pageToken != null, "pageToken must be non-null to resume a previous list() operation");
+    return listHelper(maxResults, pageToken);
+  }
+
+  /**
+   * List all items (files) and prefixes (folders) under this StorageReference.
+   *
+   * <p>This is a helper method for calling list() repeatedly until there are no more results. The
+   * default pagination size is 1000. Consistency of the result is not guaranteed if objects are
+   * inserted or removed while this operation is executing.
+   *
+   * @throws OutOfMemoryError If there are too many items at this location.
+   * @return A {@link Task} that returns all items and prefixes under the current StorageReference.
+   */
+  @NonNull
+  @PublicApi
+  public Task<ListResult> listAll() {
+    TaskCompletionSource<ListResult> pendingResult = new TaskCompletionSource<>();
+
+    List<ListResult> results = new ArrayList<>();
+    int[] prefixesCount = new int[] {0};
+    int[] itemsCount = new int[] {0};
+
+    Executor executor = StorageTaskScheduler.getInstance().getCommandPoolExecutor();
+    Task<ListResult> list = listHelper(/* maxResults= */ null, /* pageToken= */ null);
+
+    Continuation<ListResult, Task<Void>> continuation =
+        new Continuation<ListResult, Task<Void>>() {
+          @Override
+          public Task<Void> then(@NonNull Task<ListResult> currentPage) {
+            ListResult result = currentPage.getResult();
+            prefixesCount[0] += result.getPrefixes().size();
+            itemsCount[0] += result.getItems().size();
+            results.add(result);
+
+            if (result.getPageToken() != null) {
+              Task<ListResult> nextPage = listHelper(/* maxResults= */ null, result.getPageToken());
+              nextPage.continueWithTask(executor, this);
+            } else {
+              List<StorageReference> prefixes = new ArrayList<>(prefixesCount[0]);
+              List<StorageReference> items = new ArrayList<>(itemsCount[0]);
+              for (ListResult previousResult : results) {
```

Comment:
Option 6: Eliminate listAll() entirely. i.e. force the user to think about this situation. For prototyping, the user can just fetch the first 1000 and ignore the rest.

Reply:
Order is important for List API. Users often display all items in the UI, consistent ordering matters here.
In practice, random access may not be that useful, but it will be confusing if the developers run into this case. I would rather introduce less caveats.
If we really care about efficiency here, `ListAdaptor` would be the way to go. Not convinced if the complexity is worthwhile though. It also increases JAR size.

In the order of preference, 2 > 1 > 4.